text/0000-programmatic-style-injection.md

```diff
@@ -0,0 +1,158 @@
+
+The `stylesheets` property works on both light DOM and shadow DOM LWC components. This stylesheet injection mechanism complements the existing implicit template/stylesheet resolution mechanism.
+
+The LWC engine caches the static `stylesheets` value for the lifetime of the application during the component class definition. Modifying the `stylesheets` property after the component code is evaluated will not affect the stylesheet injected by this component. 
+
+For component inheritance, the LWC engine doesn't attempt to merge ancestor components stylesheets. Developers are in charge of merging (or not) the ancestors' stylesheets with their stylesheets.
+
+```js
+import { LightningElement } from 'lwc';
+import ancestorStylesheet from './ancestor.css';
+import childStylesheet from './child.css';
+
+class Ancestor extends LightningElement {
+    static stylesheets = [
+        ancestorStylesheet
+    ];
+}
+
+class Descendent extends Ancestor {
+    static stylesheets = [
+        ...super.stylesheets, 
+        childStylesheet
+    ];
+}
+```
+
+### Injection order
+
+The LWC engine injects stylesheets in the following order: first, the implicit stylesheet associated with the template and then all the stylesheets associated with the `stylesheets` property on the component class. This order allows the styles that are programmatically associated with the component to override the styles that are implicitly loaded via the template.
+
+In the following example the `<x-foo>` stylesheets will be injected in the following order: `foo.css`, `custom-a.css` and `custom-b.css`.
+
+```
+x/
+└── foo/
+    ├── custom-a.css
+    ├── custom-b.css
+    ├── foo.js
+    ├── foo.html
+    └── foo.css
+```
+
+```js
+// x/foo.js
+import { LightningElement } from 'lwc';
+import customStyleA from './custom-a.css';
+import customStyleB from './custom-b.css';
+
+export default Foo class extends LightningElement {
+    static stylesheets = [
+        customStyleA,
+        customStyleB
+    ];
+}
+```
+
+### LWC stylesheets modules
+
+Programmatically importing a stylesheet module from JavaScript using an import declaration is not a new concept introduced by this RFC. Such import is already supported by the LWC compiler. In fact, the template compiler is already using this mechanism to import the stylesheet implicitly associated with the template.
+
+However, until now, the compile-time behavior and runtime behavior for stylesheet modules have never been fully specified. This section fills this specification gap.
+
+Stylesheets modules are identifiable by their `.css` extension. This kind of module only exposes a single default export. The exported value is an opaque object, in other words, the internal implementation of the stylesheet is not exposed to developers. Keeping the exported value opaque is necessary until all the browser vendors agree on [Constructable Stylesheets](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet) implementation.

```

Comment:
This implies that someday the opaque object will be a constructable stylesheet. I'm not sure we can ever get to that future, due to light DOM scoped CSS and the need to evaluate the scope token at runtime, as well as `:host`, which varies based on the context of light vs shadow.

I would just say it's an opaque object, and developers shouldn't rely on the shape of that object as it may change. Either that, or we should make it truly opaque in a way that is only accessible to the engine.